### PR TITLE
fix #22622: Export barlines subtypes

### LINKED ISSUE #22622: We must stop using the "Offset" property to determine the default position of items
### Issue type

Engraving bug (incorrect score rendering)

### Summary 

What follows is just an explanation and a collection of the most obvious issues which arise from this. I am sure parts of this have been reported before. All of these problems come back to one single original design error: 

**The "Offset" property must be reserved for the _user's_ offset. It must _not_ be used for the default positions of the program.**

This is something which I've encountered over and over again. It has now surfaced yet again as a blocker for the new ["Center between staves" feature](https://github.com/musescore/MuseScore/issues/22175), unless I find I way to circumvent the issue. I'm logging this now as an issue of its own, so it's there for when we can get to it. It will be a big project, not only because we will need to rewrite all the default positions of all the items without using offsets, but also because we have years and years of scores that were generated this way, so we'll have to do migrations etc. 

## Explanation and examples

In Musescore, the default position of many items is determined by pre-applying an offset to them. See for instance:
image

Essentially, we have a concept of "default offset". This is a _disaster_, and it keeps causing an insane amount of engraving issues and general UX jank (which corresponds to code jank and hacks and workarounds underneath).

1. The concept of "default offset" is in itself contradictory. You offset something _with respect to a default_, or a reference point. A "default offset" is... not an offset. It is just the default position of the item on the score.
2. From a UX point of view, it is confusing and inconsistent. If I click on other items, like notes, I find their default position to have _zero_ offset. So I (correctly) interpret the "Offset" property to represent the _user's_ chosen position with respect to the default one. But when I select the hairpin, like above, I'm contradicted, and I have to guess that in that case the offset is meant _with respect to the bottom of the staff_. So which is it?
image

3. The "default offset" is meant to determine the default position. But the default position is _also_ determined by the autoplace system. Then consider this: what does a 2sp offset mean in this case? (Spoiler: nothing, it's absolutely meaningless). 
image

4. The "Offset" property is still supposed to hold the user's customized movement, so we have _one_ single variable (the offset) representing _both_ a default position _and_ a user position.  Let's say I want that hairpin to be slightly further away from the note. So I enter 3sp in the vertical offset field and... nothing happens. Why? Because the autoplace has already moved that hairpin to a point that's further than 3sp from the staff.
image

5. When mouse-dragging an item, the dragging operation has to move the item with respect to the position that the user sees on screen. But that position was calculated by autoplace. But autoplace already took an offset into account (the default one). But now I'm dragging so I'm also adding new offset. This can (and does, often) cause things to "explode" when dragging items, cause it's easy to run into infinite loops. To avoid this, we do a hack of "rebasing" the offset. "Rebasing" means that we add to the offset value the movement that had already been done by the autoplace. So, as soon as I start draggins this hairpin, the offset value jumps from 2sp to 5sp, even though I haven't even moved it yet.
image
Then, say, I want to put the hairpin back to the default position. I would imagine that that's what I'd get if I'd reset the offset.
image
But no: because we're now in a "offset rebased" state, the offset will indeed go back to 2sp, but meant as the original "2sp below the staff", which means we get this.
image

6. Notice from the previous example that, when the hairpin ends up overlapping the note, the "Minimum distance" property has suddenly been updated to an inexplicable negative value. That's not a bug, it's done intentionally. Why? Because, since the offset is also used to determine default positions, the autoplace system must take offsets into account, ending up with a situation where the offset of an item affects the autoplace of itself. Say, in the next example, I need to save space, so I move that dynamic inside the staff. I do that by dragging (i.e. applying an offset to) it. If we left "Minimum distance" as normal, the autoplace would take my offset into account and therefore "offset my offset" in order to prevent the collision with the staff, effectively disallowing me to move the dynamic inside the staff. To circumvent this problem, we hack the minimum distance and set it to a large negative value, so that the autoplace will not interpret that as a collision and let us do what we want.
image

### Steps to reproduce

Just try using the offset property on basically any item for more than 30 seconds.

### Screenshots/Screen recordings

_No response_

### MuseScore Version

Current master

### Regression

No.

### Operating system

Any

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by akhisud**

### CODE REVIEW

File: mscore/exportxml.cpp [hunk 1]
```diff
@@ -3762,7 +3789,7 @@ static int findTrackForAnnotations(int track, Segment* seg)
       }
 
 //---------------------------------------------------------
-//  repeatAtMeasureStart -- write repeats at begin of measure
+//  repeatAtMeasureStart -ite repeats at begin of measure
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: huh?

Comment (akhisud) [inline mscore/exportxml.cpp]: I wasn't able to reopen the other one

File: mscore/exportxml.cpp [hunk 2]
```diff
@@ -5123,7 +5153,7 @@ void ExportMusicXml::write(QIODevice* dev)
                         repeatAtMeasureStop(xml, m, strack, etrack, strack);
                   // note: don't use "m->repeatFlags() & Repeat::END" here, because more
                   // barline types need to be handled besides repeat end ("light-heavy")
-                  barlineRight(m);
+                  barlineRight(m,strack);
                   xml.etag();
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space after the comma

File: mscore/exportxml.cpp [hunk 3]
```diff
@@ -5067,9 +5094,12 @@ void ExportMusicXml::write(QIODevice* dev)
 
                         int sstaff = (staves > 1) ? st - strack + VOICES : 0;
                         sstaff /= VOICES;
+                        int segcount=0,empty=0;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space after the commy

File: mscore/exportxml.cpp [hunk 4]
```diff
@@ -5391,4 +5421,3 @@ void ExportMusicXml::harmony(Harmony const* const h, FretDiagram const* const fd
       }
 
 }
-
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: this change is not really needed

File: mscore/exportxml.cpp [hunk 5]
```diff
@@ -1494,7 +1520,8 @@ void ExportMusicXml::barlineRight(Measure* m)
                   xml.tagE("repeat direction=\"backward\"");
                   }
             }
-      xml.etag();
+      if(barlineAttributeSet==true)
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space between if and (, and around the ==

File: mscore/exportxml.cpp [hunk 6]
```diff
@@ -1484,8 +1505,13 @@ void ExportMusicXml::barlineRight(Measure* m)
                         }
                   }
             }
-      if (volta)
+      if (volta){
+            if(barlineAttributeSet==false){
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space between if and (, and around the ==

File: mscore/exportxml.cpp [hunk 7]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
-void ExportMusicXml::barlineRight(Measure* m)
+void ExportMusicXml::barlineRight(Measure* m, int strack)
       {
       const Measure* mmR1 = m->mmRest1(); // the multi measure rest this measure is covered by
+      bool isMMRest= mmR1->isMMRest();
       const Measure* mmRLst = mmR1->isMMRest() ? mmR1->mmRestLast() : 0; // last measure of replaced sequence of empty measures
       // note: use barlinetype as found in multi measure rest for last measure of replaced sequence
       BarLineType bst = m == mmRLst ? mmR1->endBarLineType() : m->endBarLineType();
       bool visible = m->endBarLineVisible();
+      if(!isMMRest) {
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space between if and (

**Comment by akhisud** [comment]
I'll change these. My travis doesn't detect these and builds successfully, so I can't identify them before submitting a PR

File: mscore/exportxml.cpp [hunk 8]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
 //   barlineRight -- search for and handle barline right
 //---------------------------------------------------------
 
-void ExportMusicXml::barlineRight(Measure* m)
+void ExportMusicXml::barlineRight(Measure* m, int strack)
       {
       const Measure* mmR1 = m->mmRest1(); // the multi measure rest this measure is covered by
+      bool isMMRest= mmR1->isMMRest();
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space before the =

File: mscore/exportxml.cpp [hunk 9]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+      bool barlineAttributeSet= false;
+      //xml.stag(QString("barline location=\"right\""));
       if (needBarStyle) {
             if (!visible) {
                   xml.tag("bar-style", QString("none"));
                   } else {
                   switch (bst) {
                         case BarLineType::DOUBLE:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space around the =

File: mscore/exportxml.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: I meant the loss of "wr" in that line

Comment (akhisud) [inline mscore/exportxml.cpp]: Got it. How do identify these before submitting a PR other than manually? My travis doesn't

**Comment by Jojo-Schmitz** [comment]
Please squash into one commit

File: mscore/exportxml.cpp [hunk 10]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
       const Measure* mmR1 = m->mmRest1(); // the multi measure rest this measure is covered by
+      bool isMMRest= mmR1->isMMRest();
       const Measure* mmRLst = mmR1->isMMRest() ? mmR1->mmRestLast() : 0; // last measure of replaced sequence of empty measures
       // note: use barlinetype as found in multi measure rest for last measure of replaced sequence
       BarLineType bst = m == mmRLst ? mmR1->endBarLineType() : m->endBarLineType();
       bool visible = m->endBarLineVisible();
+      if(!isMMRest) {
+            Segment* lastSegment= m->last();
+            Element* el = lastSegment->element(strack);
+            if(el && el->type()==Element::Type::BAR_LINE) {
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: spaces around ==, maybe () around the expression

File: mscore/exportxml.cpp [hunk 11]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+                  BarLine* bl =  static_cast<BarLine*>(el);
+                  if(bl && bl->customSubtype())
+                        bst= bl->barLineType();
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: soance before the =

File: mscore/exportxml.cpp [hunk 12]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
                   } else {
                   switch (bst) {
                         case BarLineType::DOUBLE:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
                               xml.tag("bar-style", QString("light-light"));
                               break;
                         case BarLineType::END_REPEAT:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp [hunk 13]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
                               xml.tag("bar-style", QString("light-heavy"));
                               break;
                         case BarLineType::BROKEN:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp [hunk 14]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
                               xml.tag("bar-style", QString("dashed"));
                               break;
                         case BarLineType::DOTTED:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: Travis is agnostig tho These. Your eyes shouldn't be ;-)

File: mscore/exportxml.cpp [hunk 15]
```diff
@@ -1484,8 +1505,13 @@ void ExportMusicXml::barlineRight(Measure* m)
                         }
                   }
             }
-      if (volta)
+      if (volta){
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space between ) and {

File: mscore/exportxml.cpp [hunk 16]
```diff
@@ -1484,8 +1505,13 @@ void ExportMusicXml::barlineRight(Measure* m)
...
+            if(barlineAttributeSet==false){
+                  xml.stag(QString("barline location=\"right\""));
+                  barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: https://musescore.org/en/developers-handbook/musescore-coding-rules#Coding-Style can help, and the qt2-musescore.xml from there can even automate this in QtCreator

**Comment by Jojo-Schmitz** [comment]
Travis is happy with it.
Would be nice to get @lvinken to give his official blessing to this PR ;-)

File: mscore/exportxml.cpp [hunk 17]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
                               break;
                         case BarLineType::DOTTED:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
                               xml.tag("bar-style", QString("dotted"));
                               break;
                         case BarLineType::END:
                         case BarLineType::END_START_REPEAT:
+                              xml.stag(QString("barline location=\"right\""));
+                              barlineAttributeSet=true;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp [hunk 18]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+void ExportMusicXml::barlineRight(Measure* m, int strack)
       {
       const Measure* mmR1 = m->mmRest1(); // the multi measure rest this measure is covered by
+      bool isMMRest = mmR1->isMMRest();
       const Measure* mmRLst = mmR1->isMMRest() ? mmR1->mmRestLast() : 0; // last measure of replaced sequence of empty measures
       // note: use barlinetype as found in multi measure rest for last measure of replaced sequence
       BarLineType bst = m == mmRLst ? mmR1->endBarLineType() : m->endBarLineType();
       bool visible = m->endBarLineVisible();
+      if (!isMMRest) {
+            Segment* lastSegment= m->last();
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space before =

File: mscore/exportxml.cpp [hunk 19]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+            Element* el = lastSegment->element(strack);
+            if (el && (el->type()==Element::Type::BAR_LINE)) {
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: space around == and maybe () around expression

File: mscore/exportxml.cpp [hunk 20]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+                  BarLine* bl =  static_cast<BarLine*>(el);
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: a scape too much for a change ;-)

File: mscore/exportxml.cpp [hunk 21]
```diff
@@ -1445,37 +1445,58 @@ void ExportMusicXml::barlineLeft(Measure* m)
...
+                  if(bl && bl->customSubtype())
+                        bst = bl->barLineType();
+                  }
+            }
       bool needBarStyle = (bst != BarLineType::NORMAL && bst != BarLineType::START_REPEAT) || !visible;
       Volta* volta = findVolta(m, false);
       if (!needBarStyle && !volta)
             return;
-      xml.stag(QString("barline location=\"right\""));
+      bool barlineAttributeSet= false;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: spaCE BEFORE 0

File: mscore/exportxml.cpp [hunk 22]
```diff
@@ -5067,9 +5094,12 @@ void ExportMusicXml::write(QIODevice* dev)
 
                         int sstaff = (staves > 1) ? st - strack + VOICES : 0;
                         sstaff /= VOICES;
+                        int segcount=0, empty=0;
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: spaces aroud = (forgot to mention earlier, sorry)

**Comment by akhisud** [comment]
@Jojo-Schmitz : Thanks for all the tips :) I'll be sure to keep a closer on formatting when I write code now

**Comment by akhisud** [comment]
@lvinken: 1) A barline at the end of a multi measure rest still doesn't get exported as its correct style (dotted, dashed etc). Problem is, a styled barline in the score shows as type:NORMAL when I check for it here : https://github.com/akhisud/MuseScore/blob/master/mscore/exportxml.cpp#L4833. Don't see why this happens. Do multimeasure rest measures store their segments (that in turn store a barline) in a different way from normal measures? 

2) Can't find XML representation rules that describe which staffs the barline spans. Is such a representation supported by XML at all?

**Comment by AntonioBL** [comment]
@akhisud : I don't know if this is what you are searching for, but regarding your second question you can have a look here:
http://forums.makemusic.com/viewtopic.php?t=1879&p=4481
http://www.musicxml.com/UserManuals/MusicXML/Content/EL-MusicXML-group-barline.htm
http://www.musicxml.com/UserManuals/MusicXML/Content/CT-MusicXML-part-group.htm
Regarding your first question, you could investigate how the barline type is read by MuseScore when painting the measure; I don't know which are the corresponding lines, but you could place inside a debugger a breakpoint inside barline.cpp at the painting function (BarLine::draw), load an example with a customized barline at the end of a multimeasure rest and going up in the backtrace chain to see how the barline type is retrieved in this case.
Ciao,
ABL

**Comment by akhisud** [comment]
@AntonioBL: Thanks, but I was looking for something a bit different. I was looking at how to represent barlines with custom presets, in a single staff (just like 'short' and 'tick' are pre-defined presets). I've explained here: https://musescore.org/en/node/22622#comment-467296.

**Comment by akhisud** [comment]
**EDIT:** 1) Multimeasure rests problem will get sorted if this PR is merged: https://github.com/musescore/MuseScore/pull/2518
2) XML doesn't seem to support span presets other than short 1 and tick 1.

So I guess, this fix is complete now.

**Comment by lasconic** [comment]
The issue is fixed on master already. See https://musescore.org/en/node/22622

**Pull request closed without merging**